overlord/devicestate/systems_test.go

```diff
@@ -227,35 +385,51 @@ func (s *createSystemSuite) TestCreateSystemFromAssertedSnaps(c *C) {
 		return nil
 	}
 
-	dir, err := devicestate.CreateSystemForModelFromValidatedSnaps(model, "1234", s.db, infoGetter, snapWriteObserver)
+	dir, err := devicestate.CreateSystemForModelFromValidatedSnaps(s.state, model, "1234", s.db, &infoGetter, snapWriteObserver)
 	c.Assert(err, IsNil)
 	c.Check(newFiles, DeepEquals, []string{
 		filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps/snapd_4.snap"),
-		filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps/pc-kernel_1.snap"),
+		filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps/pc-kernel-with-kmods_1.snap"),
+		filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps/pc-kernel-with-kmods+kmod_11.comp"),
 		filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps/core20_3.snap"),
 		filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps/pc_2.snap"),
 		filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps/other-present_5.snap"),
 		filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps/other-required_6.snap"),
 		filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps/other-core18_7.snap"),
 		filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps/core18_8.snap"),
+		filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps/snap-with-components_2.snap"),
+		filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps/snap-with-components+comp-1_22.comp"),
 	})
 	c.Check(dir, Equals, expectedDir)
+
 	// naive check for files being present
 	for _, info := range infos {
 		c.Check(filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps", filepath.Base(info.MountFile())),
 			testutil.FileEquals,
 			testutil.FileContentRef(info.MountFile()))
 	}
+	for _, compInfo := range compInfos {
+		cpi := snap.MinimalComponentContainerPlaceInfo(
+			compInfo.Component.ComponentName,
+			compInfo.Revision,
+			compInfo.Component.SnapName,
+		)
+
+		c.Check(filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps", filepath.Base(cpi.MountFile())),
+			testutil.FileEquals,
+			testutil.FileContentRef(cpi.MountFile()))
+	}
+
 	// recovery system bootenv was set
 	c.Check(bl.RecoverySystemDir, Equals, "/systems/1234")
 	c.Check(bl.RecoverySystemBootVars, DeepEquals, map[string]string{
 		"snapd_full_cmdline_args":  "mock static args from gadget",
 		"snapd_extra_cmdline_args": "",
-		"snapd_recovery_kernel":    "/snaps/pc-kernel_1.snap",
+		"snapd_recovery_kernel":    "/snaps/pc-kernel-with-kmods_1.snap",
 	})
 	// load the seed
 	validateCore20Seed(c, "1234", model, s.storeSigning.Trusted,
-		"other-core18", "core18", "other-present", "other-required")
+		"other-core18", "core18", "other-present", "other-required", "snap-with-components")
 }
 
 func (s *createSystemSuite) TestCreateSystemFromUnassertedSnaps(c *C) {
```

Comment:
this PR is both a refactor plus a bit of initial component code, because of the refactor part it would be good if the test changes were purely additive

Reply:
Fixed this, I ended up rebasing so the changes are part of the original commits.